### PR TITLE
RFC 98: Remote channels for cross-browsing-group communication

### PR DESCRIPTION
**Pull request by jgraham**
[Rendered](https://github.com/web-platform-tests/rfcs/blob/remote_channel/rfcs/remote_channel.md)
[Implementation](https://github.com/web-platform-tests/wpt/pull/29803)

### CODE REVIEW

File: rfcs/remote_channel.md
Comment (Ms2ger) [inline rfcs/remote_channel.md:40]:
```suggestion
  [SpecialPowers.spawn](https://searchfox.org/mozilla-central/source/testing/specialpowers/content/SpecialPowersChild.jsm#1547-1584)
```

File: rfcs/remote_channel.md
Comment (Ms2ger) [inline rfcs/remote_channel.md:104]:
```suggestion
here is that they are designed to allow blocking reads. This means we
```

File: rfcs/remote_channel.md
Comment (Ms2ger) [inline rfcs/remote_channel.md:491]:
```suggestion
general approach is to explicitly store a refcount as part of the
```

File: rfcs/remote_channel.md
Comment (Ms2ger) [inline rfcs/remote_channel.md:523]:
```suggestion
  let remote = new RemoteWindow();
```

File: rfcs/remote_channel.md
Comment (Ms2ger) [inline rfcs/remote_channel.md:615]:
```suggestion
<!--  LocalWords:  testdriver websockets bfcache
```

**Comment by jgraham** [comment]
Gentle ping here. This has been open for a week, so per the process I could go ahead and merge it, but I'm pretty sure you don't want me to do that ;)

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:249]: Based on my experience on bfcache tests, I feel `executeScriptNoResult()` might be hard to use, because:

- Mixing `await executeScriptNoResult()` and `await executeScript()` might complicates the timeline. Having `await executeScript()` only would be clear, because it serializes the script evaluation on caller and remote context.
- The choice between `await executeScriptNoResult()` and `await executeScript()` might depend on internals of `executeScript()`. When I was migrating my bfcache tests from COOP-based framework to test_driver-based framework, I had to make changes around `executeScriptNoResult()`-equivalent that were hard to explain without knowing the internal structure of the framework, so I switched to the `executeScript()`-only interface.

So I'd prefer removing `executeScriptNoResult()`.
Still there would remain complexities around navigation: instead of `executeScriptNoResult()`, I created promises / async callbacks triggered by `executeScript()` but not blocking `async executeScript()` in bfcache tests. But by removing `executeScriptNoResult()` I expect we can make the complexities only navigation-specific, not affecting `RemoteWindow` interface.

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:267]: As commented at https://github.com/web-platform-tests/wpt/pull/29803#discussion_r707551472, I'm not sure how `pause` works and whether `pause` is needed (the draft impl doesn't use `pause`).

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:71]: nit: Maybe it would be clearer to move these blocks to "Prior Art" section, because the list here isn't a part of the proposal (e.g. this RFC doesn't propose SpecialPowers-integrated API).

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:152]: nit: It might be better to clarify that this serialization is implemented by pure JavaScript (IIUC `RemoteObject.serialize()` in the draft PR) and thus doesn't depend on WebDriver BiDi impl itself.

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:69]: I'd like to clarify (somewhere in the RFC) what are the dependencies of this RFC in terms of implementation.

IIUC this RFC can be implementable by pure JavaScript outside WPT infra + existing server-side stash, which is good because we can run the tests by running wpt server and bare browsers, without further depending on WPT test infra.
This RFC mentions integration with WebDriver BiDi, test_driver and testharness, but these are out of scope of this RFC itself.

Is my understanding correct?
Why the changes to `resources/testharness.js` and `tools/wptrunner/wptrunner` are needed in https://github.com/web-platform-tests/wpt/pull/29803 ?

(Other changes, in JS/HTML files, `tools/wptserve/wptserve/stash.py`, and `websockets/handlers/msg_channel_wsh.py` look like "implementable by pure JavaScript outside WPT infra + existing server-side stash")

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:19]: Is this RFC going to replace the framework in the COOP/COEP tests?

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:256]: Doesn't `RemoteContext` work (as proposed in https://github.com/web-platform-tests/rfcs/pull/91)?

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:574]: Would `RemoteContext.testdriver.click` be a pure JavaScript wrapper like `RemoteContext.executeScript(() => test_driver.click())`?

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:568]: For further integration with WebDriver, testdriver etc., discussion about benefits and risks would be helpful (not necessarily now though, given that this is "Possible Future Additions").

One one hand ("Risk" side), I basically prefer not integrating the framework too much to WebDriver/test_driver etc., to make the framework work with minimum dependencies and work with bare browsers, and not to make debugging harder. We might want to `JSON.stringify()` instead of WebDriver BiDi format if we want to reduce complexity around serialization, given that the current tests considered so far don't need WebDriver BiDi capability beyond `JSON.stringify()`.
So I'm interested in hearing more about the benefits (from the point of view of test writers and WPT infra maintainers) when we consider the further integration.

File: rfcs/remote_channel.md
Comment (hiroshige-g) [inline rfcs/remote_channel.md:258]: I did a similar thing to `start_window()` in my bfcache tests and I was also not sure about the better naming, better semantics, etc. (and thus I just made it `const executor = new Executor(uuid);` and avoided explicit naming).
So I'm also curious how this should be named.

**Comment by hiroshige-g** [comment]
Commented somehow verbosely to move discussion forward.

As for `executeScript()`, the RFC basically looks similar to what the current `RemoteContext.execute_script()` and bfcache tests do, while there are more primitives around navigations (`executeScriptNoResult`, `pause` and `closeAllChannelSockets` in this RFC while the bfcache tests only have `closeAllChannelSockets`-equivalent, as commented inline above) and I'd like to understand the differences (IIUC there aren't fundamental differences, because I expect switching the impl from Fetch API to websocket doesn't impact so much).
So in terms of `executeScript()`, this RFC looks like switching from Fetch API to websocket based implementation (compared to the previous RFCs and https://github.com/web-platform-tests/wpt/pull/28950), both pure JavaScript-based, and thus doesn't cause test bodies except for mechanical changes (correct?).

The RFC is more about replacing the existing `send()` and `receive()` primitives currently used in COEP/COOP tests, by introducing related API classses, adding more APIs in addition to `send()`/`receive()` etc., and I expect test writers (other than me) and WPT infra people might have more interests on these.

cc/ @ArthurSonzogni and @foolip.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:249]: So (for the context of others) the technical concern here is that for bfcache tests in particular you need all sockets to be closed before initiating the navigation. The problem is that if you send a message, execute some script that closes all sockets, and then send the result, you end up re-opening a socket for the result. So there are broadly two choices:
* Don't try to send a result at all. `postMessage` works like this, and `executeScriptNoResult` does the same: we simply don't provide a channel to send the result on so there's no concern about sockets being reopened.
* Send a result, but ensure we navigate after the result has been sent, so nothing will implictly reopen a socket. Practically this means something like `setTimeout(async url => {await closeAllSockets(); startNavigation(url)}, 0)`. The problem with that is that we are currently linting for `setTimeout` so the infra discourages even usage that's required like this.

I don't quite get the argument about serialization; in the case that you're putting the actual action of the remote script in a timeout the test window can guarantee the script has been sent, but the actual action of the script hasn't happened at the time the value is returned, so it doesn't seem like that method offers particuarly stronger guarantees around ordering. I suppose one advantage is that any exception that happens when evalutaing the script is reported through the return value, rather than just happening in the remote window.

So, I think I'm mostly OK with dropping `executeScriptNoResult` because, as you say, it's exposing an implementation detail and it seems cleaner to start with a smaller API. My concern is that for some kinds of tests (notably bfcache), understanding that implementation detail is very important, and there's an argument that making it an explicit part of the API will help people write correct tests. So I'd like to hear other opinions here.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:267]: So the problem pause is intended to solve is where you have something like:
```
await remote.executeScript(() => startNavigation());
await remote.executeScript(() => ensureNavigationHappened());
```
With a `poll()` based interface this can work as written; the remote simply doesn't call `poll()` after `startNavigation` and before the navigation is complete. But with websockets, we don't have any synchronisation between the navigation and sending the `ensureNavigationHappened` message. If the test sends the second message before the socket is closed on the remote, it can end up being lost. So pause solves this by effectively providing a synchronisation point around the socket being closed:
```
await remote.executeScript(() => startNavigation());
remote.pause();
// We now know the remote now won't listen for messages until it reconnects to the socket after navigation
await remote.executeScript(() => ensureNavigationHappened());
```

From the point of view of the remote, `closeAllSockets()` and `pause()` do race, but the effect of `pause()` is a server-initiated disconnect of the read channel for the remote, and the effect of `closeAllSockets` is a client-initiated disconnect of all channel-related sockets, so the ordering isn't very important.

We can avoid requiring pause by not sending requests to a remote after initiating a navigation, before getting a confirmation that the navigation is actually complete. In practice this means that the target page has to be a different `RemoteWindow` (i.e. have a different uuid) and has to process a message i.e. something like:
```
await remote.executeScript(() => startNavigation());
await navigationTargetRemote.executeScript(() => goBack());
await remote.executeScript(() => ensureNavigationHappened());
```
That's what the bfcache tests are doing now, and why `pause` isn't needed there, but either way the ergonomics are a bit tricky. I think we could drop it from the initial featureset, but these races around navigation are tricky to understand and debug.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:19]: I don't expect the RFC to require moving the COOP/COEP tests. However I expect the RFC to provide all the features needed for those tests, and would anticipate the actual migration happening as a followup once the implementation has landed.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:256]: No, because `RemoteContext` is already used by testharness.js to mean something slightly different:
```
    /*
     * A RemoteContext listens for test events from a remote test context, such
     * as another window or a worker. These events are then used to construct
     * and maintain RemoteTest objects that mirror the tests running in the
     * remote context.
     */
```

In the long term it would make sense to unify these features, but in the short term a naming conflict will cause breakage.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:574]: That doesn't work as written, because all testdriver commands have to go via the test window since that's the one that webdriver is using. So it would desugar into `testdriver.click(RemoteContext.uuid)`, and testdriver would learn to look up contexts from a `uuid` parameter in the URL, as well as from the internal identifier it uses today.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:568]: Yes, future additions here would be subject to the same RFC process. So the idea was just to sketch out some possibilites in case people felt there were specific parts that should be given additional consideration in the initial design, or would be high value to work on sooner.

In terms of the serialization format, we are already using extra complexity beyond JSON-stringify when serializing a `SendChannel` for message responses. We obviously could special case that, but in general I think using plain JSON as a serialization format for data where we might want to support non-JSON types is a bad idea (e.g. current WebDriver mostly uses plain JSON but also has super-hacky support for serializing Element objects which doesn't generalise to other types and is generally a mess). Aiming for a featureset closer to structured clone seems more future proof, but structured clone doesn't actually define a serialization. So, given that WebDriver BiDi has very similar requirements, and actually defines the wire format, it seems sensible to use that as a reference point. The fact that, going forward, we could use WebDriver BiDi instead of the stash as the backend in the wptrunner implementation is good, but it's not the primary motivation here.

**Comment by jgraham** [comment]
@hiroshige-g Thanks for the detailed comments, it's much appreciated! I've responded to the technical concerns, hopefully in a way that also provides context for other reviewers. I'll update the RFC for the more editorial issues. Please let e know if I miss[ed] anything.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:152]: I've clarified this below, let me know if you think more is required.

**Comment by jgraham** [comment]
Ping, again. This has now gone another week since the last round of updates without further review.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md:669]: Do you want to keep them, then?

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md]: I see, I thought it might silently work.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md:110]: I didn't realize it was a question about how `Queue` works, but I guess it was. I assumed reads were blocking, but it has arguments, and you're using `queue.get(True, 1)` in a loop, so the problem I imagined doesn't exist.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md]: Just `call` seems OK.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md:271]: There's a similar thing that happens in BiDi where I think it could be distinguished, but not here, I think you're right.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md:276]: I see. I think I must have misread this too, thinking that an `exceptionDetails` property would be added to the re-thrown error, but this is talking about a response.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md]: Resolving this, you renamed it `disconnectReader`.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md]: With `type` added I think it's EventTarget-like enough for nobody to notice the difference.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md:545]: Well, I think the scheme you have here shouldn't leak more and more data as the tests run, so that's the most important thing. If the only scenario where the queue might be deleted too soon is when one deliberately reuses a UUID for a second `RecvChannel`, then that's not worth trying to fix.

Honestly I don't trust that I can see all the potential issues here, better to think about that when reviewing the implementation.

File: rfcs/remote_channel.md
Comment (jgraham) [reply rfcs/remote_channel.md:669]: Well it's convenient for me when editing, but I could remove them once the overall RFC is approved, right before merging.

File: rfcs/remote_channel.md
Comment (foolip) [reply rfcs/remote_channel.md]: Alright, enough bikeshedding :)

File: rfcs/remote_channel.md
Comment (foolip) [inline rfcs/remote_channel.md:242]: Good clarification!

File: rfcs/remote_channel.md
Comment (foolip) [inline rfcs/remote_channel.md:298]:
```suggestion
  initiate a connection to the top-level test context. For example,
```

**Review by foolip — APPROVED** [review]
Just some missing punctuation.

**Pull request merged**
